app/controllers/LongerCommodityCodeController.scala

```diff
@@ -53,79 +52,77 @@ class LongerCommodityCodeController @Inject() (
     extends FrontendBaseController
     with I18nSupport {
 
-  private val form = formProvider()
+  private val form           = formProvider()
+  private val validLength    = 6
+  private val maxValidLength = 10
 
   def onPageLoad(mode: Mode, recordId: String): Action[AnyContent] = (identify andThen getData andThen requireData) {
     implicit request =>
-      val categorisationInfoOpt: Option[CategorisationInfo] =
-        request.userAnswers.get(RecordCategorisationsQuery).flatMap(_.records.get(recordId))
+      val shortComcodeOpt = getShortCommodityCodeOpt(recordId, request, validLength)
 
-      val originalComcodeOpt = categorisationInfoOpt.flatMap(_.originalCommodityCode)
-      val latestComcodeOpt   = categorisationInfoOpt.map(_.commodityCode)
-
-      val answerToFillFormWith = for {
-        originalCode <- originalComcodeOpt
-        latestCode   <- latestComcodeOpt
-      } yield latestCode.drop(originalCode.length)
-
-      val preparedForm = (categorisationInfoOpt, answerToFillFormWith) match {
-        case (Some(categorisationInfo), Some(answerToFillFormWith)) if categorisationInfo.latestDoesNotMatchOriginal =>
-          form.fill(answerToFillFormWith)
-        case _                                                                                                       =>
-          form
+      val preparedForm = request.userAnswers.get(LongerCommodityCodePage(recordId)) match {
+        case None        => form
+        case Some(value) => form.fill(value)
       }
 
-      originalComcodeOpt match {
-        case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
-          Ok(view(preparedForm, mode, commodityCodeSansTrailingZeros(shortCommodity), recordId))
-        case _                                                                                  =>
+      shortComcodeOpt match {
+        case Some(shortComcode) if shortComcode.length == validLength =>
+          Ok(view(preparedForm, mode, shortComcode, recordId))
+        case _                                                        =>
           Redirect(routes.JourneyRecoveryController.onPageLoad().url)
       }
   }
 
   def onSubmit(mode: Mode, recordId: String): Action[AnyContent] =
     (identify andThen getData andThen requireData).async { implicit request =>
-      val categorisationInfoOpt = request.userAnswers.get(RecordCategorisationsQuery).flatMap(_.records.get(recordId))
-
-      val originalComcodeOpt = categorisationInfoOpt.flatMap(_.originalCommodityCode)
-      val latestComcodeOpt   = categorisationInfoOpt.map(_.commodityCode)
-
-      originalComcodeOpt match {
-        case Some(shortCommodity) if commodityCodeSansTrailingZeros(shortCommodity).length == 6 =>
-          val shortCode = commodityCodeSansTrailingZeros(shortCommodity)
+      val shortComcodeOpt         = getShortCommodityCodeOpt(recordId, request, validLength)
+      val currentlyCategorisedOpt =
```

Comment:
You have a few options floating about that you use lower down in the code. To make it more functional it might be worth refactoring so you don't have these gaps and instead transform the data to what you wish to return

Tbf it looks like it works so I don't care if it stays like this lol